File: typescript/src/serializer.ts
// Reflex — Workflow Serializer
// Implements M7-3: Serialize programmatic Workflow objects to JSON

import type {
  BlackboardReader,
  BuiltinGuard,
  CustomGuard,
  Edge,
  Guard,
  Node,
  Workflow,
} from './types.js';
import { WorkflowValidationError } from './registry.js';

// ---------------------------------------------------------------------------
// Public types
// ---------------------------------------------------------------------------

/** Maps evaluate functions back to guard names for serialization. */
export type GuardNameMap = Map<
  (blackboard: BlackboardReader) => boolean,
  string
>;

/** Options for serializeWorkflow. */
export interface SerializeWorkflowOptions {
  guardNames?: GuardNameMap;
}

// ---------------------------------------------------------------------------
// serializeWorkflow
// ---------------------------------------------------------------------------

/**
 * Serialize a typed Workflow object to a JSON string.
 *
 * BuiltinGuards serialize directly. CustomGuards require a `guardNames` map
 * to recover the JSON `{ type: "custom", name: "..." }` representation.
 *
 * @param wf - The Workflow to serialize
 * @param options - Optional guard name map for custom guard serialization
 * @returns A JSON string representing the workflow
 * @throws WorkflowValidationError if a custom guard is not in guardNames
 */
export function serializeWorkflow(
  wf: Workflow,
  options?: SerializeWorkflowOptions,
): string {
  const out: Record<string, unknown> = {
    id: wf.id,
    entry: wf.entry,
    nodes: serializeNodes(wf.nodes),
    edges: serializeEdges(wf.edges, wf.id, options?.guardNames),
  };

  if (wf.metadata !== undefined) {
    out.metadata = wf.metadata;
  }

  return JSON.stringify(out, null, 2);
}

// ---------------------------------------------------------------------------
// Internal helpers
// ---------------------------------------------------------------------------

function serializeNodes(
  nodes: Record<string, Node>,
): Record<string, Record<string, unknown>> {
  const out: Record<string, Record<string, unknown>> = {};
  for (const [key, node] of Object.entries(nodes)) {
    const n: Record<string, unknown> = {
      id: node.id,
      spec: node.spec,
    };
    if (node.description !== undefined) {
      n.description = node.description;
    }
    if (node.invokes !== undefined) {
      n.invokes = {
        workflowId: node.invokes.workflowId,
        returnMap: node.invokes.returnMap.map((rm) => ({
          parentKey: rm.parentKey,
          childKey: rm.childKey,
        })),
      };
    }
    out[key] = n;
  }
  return out;
}

function serializeEdges(
  edges: Edge[],
  workflowId: string,
  guardNames?: GuardNameMap,
): Array<Record<string, unknown>> {
  return edges.map((edge) => {
    const e: Record<string, unknown> = {
      id: edge.id,
      from: edge.from,
      to: edge.to,
      event: edge.event,
    };
    if (edge.guard !== undefined) {
      e.guard = serializeGuard(edge.guard, workflowId, edge.id, guardNames);
    }
    return e;
  });
}

function serializeGuard(
  guard: Guard,
  workflowId: string,
  edgeId: string,
  guardNames?: GuardNameMap,
): Record<string, unknown> {
  if (guard.type === 'custom') {
    const cg = guard as CustomGuard;
    const name = guardNames?.get(cg.evaluate);
    if (!name) {
      throw new WorkflowValidationError(
        'SCHEMA_VIOLATION',
        workflowId,
        `Edge '${edgeId}': custom guard has no name in guardNames map (cannot serialize)`,
        { edgeId },
      );
    }
    return { type: 'custom', name };
  }

  // Builtin guard
  const bg = guard as BuiltinGuard;
  const out: Record<string, unknown> = { type: bg.type, key: bg.key };
  if (bg.value !== undefined) {
    out.value = bg.value;
  }
  return out;
}

File: go/serializer.go
package reflex

import (
	"encoding/json"
	"fmt"
)

// ---------------------------------------------------------------------------
// Public types
// ---------------------------------------------------------------------------

// GuardNameMap maps Guard implementations back to their JSON names for
// serialization. This is the inverse of GuardRegistry used by LoadWorkflow.
type GuardNameMap map[Guard]string

// SerializeWorkflowOptions configures optional parameters for SerializeWorkflow.
type SerializeWorkflowOptions struct {
	GuardNames GuardNameMap
}

// ---------------------------------------------------------------------------
// Intermediate JSON types (for edges with guards)
// ---------------------------------------------------------------------------

type jsonGuardOut struct {
	Type  string `json:"type"`
	Key   string `json:"key,omitempty"`
	Value any    `json:"value,omitempty"`
	Name  string `json:"name,omitempty"`
}

type jsonEdgeOut struct {
	ID    string        `json:"id"`
	From  string        `json:"from"`
	To    string        `json:"to"`
	Event string        `json:"event"`
	Guard *jsonGuardOut `json:"guard,omitempty"`
}

// ---------------------------------------------------------------------------
// SerializeWorkflow
// ---------------------------------------------------------------------------

// SerializeWorkflow produces a JSON representation of a Workflow.
// BuiltinGuards serialize directly. CustomGuardFuncs require a GuardNameMap
// in opts to recover the JSON { type: "custom", name: "..." } representation.
func SerializeWorkflow(wf *Workflow, opts *SerializeWorkflowOptions) ([]byte, error) {
	edges := make([]jsonEdgeOut, len(wf.Edges))
	for i, e := range wf.Edges {
		je := jsonEdgeOut{
			ID:    e.ID,
			From:  e.From,
			To:    e.To,
			Event: e.Event,
		}
		if e.Guard != nil {
			g, err := serializeGuard(e.Guard, wf.ID, e.ID, opts)
			if err != nil {
				return nil, err
			}
			je.Guard = g
		}
		edges[i] = je
	}

	out := struct {
		ID       string           `json:"id"`
		Entry    string           `json:"entry"`
		Nodes    map[string]*Node `json:"nodes"`
		Edges    []jsonEdgeOut    `json:"edges"`
		Metadata map[string]any   `json:"metadata,omitempty"`
	}{
		ID:       wf.ID,
		Entry:    wf.Entry,
		Nodes:    wf.Nodes,
		Edges:    edges,
		Metadata: wf.Metadata,
	}

	return json.MarshalIndent(out, "", "  ")
}

// ---------------------------------------------------------------------------
// Guard serialization
// ---------------------------------------------------------------------------

func serializeGuard(g Guard, wfID, edgeID string, opts *SerializeWorkflowOptions) (*jsonGuardOut, error) {
	switch guard := g.(type) {
	case *BuiltinGuard:
		out := &jsonGuardOut{
			Type: string(guard.Type),
			Key:  guard.Key,
		}
		if guard.Type == GuardEquals || guard.Type == GuardNotEquals {
			out.Value = guard.Value
		}
		return out, nil

	case *CustomGuardFunc:
		if opts == nil || opts.GuardNames == nil {
			return nil, newValidationError(ErrSchemaViolation, wfID,
				fmt.Sprintf("edge '%s': custom guard has no name in GuardNames map (cannot serialize)", edgeID))
		}
		name, ok := opts.GuardNames[g]
		if !ok {
			return nil, newValidationError(ErrSchemaViolation, wfID,
				fmt.Sprintf("edge '%s': custom guard has no name in GuardNames map (cannot serialize)", edgeID))
		}
		return &jsonGuardOut{Type: "custom", Name: name}, nil

	default:
		return nil, newValidationError(ErrSchemaViolation, wfID,
			fmt.Sprintf("edge '%s': unknown guard type %T (cannot serialize)", edgeID, g))
	}
}
